st3/lsp_utils/_client_handler/abstract_plugin.py

```diff
@@ -22,35 +24,52 @@
     'scopes': Optional[List[str]],
     'syntaxes': Optional[List[str]],
 }, total=False)
+ApiNotificationHandler = Callable[[Any], None]
+ApiRequestHandler = Callable[[Any, Callable[[Any], None]], None]
 
 
 class ApiWrapper(ApiWrapperInterface):
-    def __init__(self, plugin: AbstractPlugin):
+    def __init__(self, plugin: 'ref[AbstractPlugin]'):
         self.__plugin = plugin
 
+    def __session(self):
+        plugin = self.__plugin()
+        return plugin.weaksession() if plugin else None
+
     # --- ApiWrapperInterface -----------------------------------------------------------------------------------------
 
-    def on_notification(self, method: str, handler: Callable[[Any], None]) -> None:
-        setattr(self.__plugin, method2attr(method), lambda params: handler(params))
+    def on_notification(self, method: str, handler: ApiNotificationHandler) -> None:
+        def handle_notification(handler_ref: 'ref[ApiNotificationHandler]', params: Any) -> None:
+            handler = handler_ref()
+            if handler:
+                handler(params)
+
+        plugin = self.__plugin()
+        if plugin:
+            setattr(plugin, method2attr(method), partial(ref(handler), handle_notification))
 
-    def on_request(self, method: str, handler: Callable[[Any, Callable[[Any], None]], None]) -> None:
+    def on_request(self, method: str, handler: ApiRequestHandler) -> None:
         def send_response(request_id: Any, result: Any) -> None:
-            session = self.__plugin.weaksession()
+            session = self.__session()
             if session:
                 session.send_response(Response(request_id, result))
 
-        def on_response(params: Any, request_id: Any) -> None:
-            handler(params, lambda result: send_response(request_id, result))
+        def on_response(handler_ref: 'ref[ApiRequestHandler]', params: Any, request_id: Any) -> None:
+            handler = handler_ref()
+            if handler:
```

Comment:
Hmm, the reference is dead here. Must be missing something. Why would it go dead on me if the original `handler` method (the plugin instance method) is not dead...?

Reply:
Fixed with c5fbaaeee94e8d8c99961f36039f72bdb71456bb